rflx/pyrflx/typevalue.py

```diff
@@ -922,57 +931,50 @@ def bytestring(self) -> bytes:
     def fields(self) -> List[str]:
         return [f.name for f in self._type.fields]
 
-    def _update_accessible_fields(self) -> None:
-        nxt = self._next_field(INITIAL.name)
-        fields: List[str] = []
-        while nxt and nxt != FINAL.name:
-
-            if (
-                self.__simplified(self._type.field_condition(Field(nxt))) != TRUE
-                or not self._has_first(nxt)
-                or (
-                    not self._has_length(nxt)
-                    if not isinstance(self._fields[nxt].typeval, OpaqueValue)
-                    else not self._is_valid_opaque_field(nxt)
-                )
-            ):
-                break
-
-            fields.append(nxt)
-            nxt = self._next_field(nxt)
-        self.accessible_fields = fields
-
     def _is_valid_opaque_field(self, field: str) -> bool:
-        if self._get_length_unchecked(field) == UNDEFINED:
-            return False
 
-        for edge in self._type.incoming(Field(field)):
+        typeval = self._fields[field].typeval
+        assert isinstance(typeval, (ScalarValue, CompositeValue))
+        incoming = self._type.incoming(Field(field))
+        if isinstance(typeval, CompositeValue):
+            for l in incoming:
+                if (
+                    self.__simplified(l.condition) == TRUE
+                    and l.length != UNDEFINED
+                    and self._fields[l.source.name].set
+                ):
+                    break
+            else:
+                return False
+
+        for edge in incoming:
             if self.__simplified(edge.condition) == TRUE:
                 valid_edge = edge
                 break
         else:
             return True
```

Comment:
The second for-loop is unnecessary and it doesn't make sense to support scalar fields in this function.

```diff
diff --git a/rflx/pyrflx/typevalue.py b/rflx/pyrflx/typevalue.py
index 6bd632e9..62c85a98 100644
--- a/rflx/pyrflx/typevalue.py
+++ b/rflx/pyrflx/typevalue.py
@@ -932,27 +932,18 @@ class MessageValue(TypeValue):
         return [f.name for f in self._type.fields]
 
     def _is_valid_opaque_field(self, field: str) -> bool:
-
-        typeval = self._fields[field].typeval
-        assert isinstance(typeval, (ScalarValue, CompositeValue))
+        assert isinstance(self._fields[field].typeval, CompositeValue)
         incoming = self._type.incoming(Field(field))
-        if isinstance(typeval, CompositeValue):
-            for l in incoming:
-                if (
-                    self.__simplified(l.condition) == TRUE
-                    and l.length != UNDEFINED
-                    and self._fields[l.source.name].set
-                ):
-                    break
-            else:
-                return False
-
-        for edge in incoming:
-            if self.__simplified(edge.condition) == TRUE:
-                valid_edge = edge
+        for l in incoming:
+            if (
+                self.__simplified(l.condition) == TRUE
+                and l.length != UNDEFINED
+                and self._fields[l.source.name].set
+            ):
+                valid_edge = l
                 break
         else:
-            return True
+            return False
 
         return all(
             (v.name in self._fields and self._fields[v.name].set) or v.name == "Message"
diff --git a/tests/test_pyrflx.py b/tests/test_pyrflx.py
index 4b87baa6..4dc9c9d2 100644
--- a/tests/test_pyrflx.py
+++ b/tests/test_pyrflx.py
@@ -795,12 +795,18 @@ def test_field_set() -> None:
 
 def test_check_nodes_opaque(tlv: MessageValue, frame: MessageValue) -> None:
     # pylint: disable=protected-access
-    assert tlv._is_valid_opaque_field("Length")
     assert not tlv._is_valid_opaque_field("Value")
+    tlv.set("Tag", "Msg_Data")
+    tlv.set("Length", 1000)
+    assert tlv._is_valid_opaque_field("Value")
+
     frame.set("Destination", 2 ** 48 - 1)
     frame.set("Source", 0)
     frame.set("Type_Length_TPID", 1501)
     assert not frame._is_valid_opaque_field("Payload")
+    frame.set("Type_Length_TPID", 1500)
+    frame.set("Type_Length", 1500)
+    assert frame._is_valid_opaque_field("Payload")
 
 
 def test_icmp_parse_binary(icmp: MessageValue) -> None:
```

Reply:
Fixed in https://github.com/Componolit/RecordFlux/pull/381/commits/43cb59e028e040be26d5e714d897852fde513ae0.